### PR TITLE
refactor(tracing): simplified the dynamic hooks code

### PR DESCRIPTION
**Pull request by bungle**
### Summary

Removes ugly code generation with simple local functions.

KAG-3870

### CODE REVIEW

**Comment by bungle** [comment]
The non-varargs optimization can be added if neccessary.

File: kong/dynamic_hook/init.lua [hunk 1]
```diff
@@ -23,39 +29,123 @@ local non_function_hooks = {
...
+    if should_execute_original_func(group_name) then
+      return original_func(...)
+    end
+    execute_hooks_vararg(handlers.befores, "before", group_name, ...)
+    return execute_after_hooks_vararg(handlers, group_name, original_func(...))
+  end
+end
+
+
+local function execute_hook(hook, hook_type, group_name, a1, a2, a3, a4, a5, a6, a7, a8)
```
Comment (ADD-SP) [inline kong/dynamic_hook/init.lua:97]: Some functions (such as cosocket) has strict checks of the number of arguments, so we cannot do something like `a0, a1, a2, a3 ......`.

**Review by ADD-SP — CHANGES_REQUESTED** [review]

**Review by ADD-SP — COMMENTED** [review]
Since we need to profile and benchmark this PR, could you please convert this PR as a draft?

File: kong/dynamic_hook/init.lua [hunk 2]
```diff
@@ -93,12 +184,12 @@ function _M.is_group_enabled(group_name)
 end
 
 
-function _M.run_hooks(ctx, group_name, hook_name, ...)
+function _M.run_hooks(_, group_name, hook_name, ...)
```
[the remark this answers is not in the document]
Comment (bungle) [reply kong/dynamic_hook/init.lua:188]: @ADD-SP it is not used anywhere.

File: kong/dynamic_hook/init.lua (hunk 1 above)
Comment (bungle) [reply kong/dynamic_hook/init.lua:97]: Perhaps this problem should be handled on hook side.

**Comment by bungle** [comment]
> Since we need to profile and benchmark this PR, could you please convert this PR as a draft?

Converted to draft.

File: kong/dynamic_hook/init.lua (hunk 1 above)
Comment (bungle) [reply kong/dynamic_hook/init.lua:97]: I made a workaround to C code thing.

Comment (ADD-SP) [reply kong/dynamic_hook/init.lua:97]: > @ADD-SP you mean those that are implemented in C? (in Lua land it most likely doesn't matter)

Yes, the Lua C function might have strict checks on the number of arguments.

**Comment by ADD-SP** [comment]
Rebased on master.

File: kong/dynamic_hook/init.lua [hunk 3]
```diff
@@ -110,28 +226,27 @@ function _M.run_hooks(ctx, group_name, hook_name, ...)
 
   local ok, err = pcall(handler, ...)
```
Comment (ADD-SP) [inline kong/dynamic_hook/init.lua:227]: `...` is NYI here, can we replace it with `a0, a1, a2, a3` and set up safeguards to handle more than N arguments?

```
---- TRACE 1434 start 1422/0 @./kong/dynamic_hook/init.lua:194
0001  UGET     1   0      ; ALWAYS_ENABLED_GROUPS       (@./kong/dynamic_hook/init.lua:194)
0002  TGETV    1   1   0       (@./kong/dynamic_hook/init.lua:194)
0003  ISF          1       (@./kong/dynamic_hook/init.lua:194)
0004  JMP      2 => 0007
0005  KPRI     1   2       (@./kong/dynamic_hook/init.lua:195)
0006  RET1     1   2       (@./kong/dynamic_hook/init.lua:195)
0005  IST          3       (@./kong/dynamic_hook/init.lua:213)
0006  JMP      3 => 0008
0008  UGET     3   1      ; NON_FUNCTION_HOOKS       (@./kong/dynamic_hook/init.lua:217)
0009  TGETV    3   3   1       (@./kong/dynamic_hook/init.lua:217)
0010  IST          3       (@./kong/dynamic_hook/init.lua:218)
0011  JMP      4 => 0013
0013  TGETV    4   3   2       (@./kong/dynamic_hook/init.lua:222)
0014  IST          4       (@./kong/dynamic_hook/init.lua:223)
0015  JMP      5 => 0017
0017  UGET     5   2      ; pcall       (@./kong/dynamic_hook/init.lua:227)
0018  MOV      7   4       (@./kong/dynamic_hook/init.lua:227)
0019  VARG     8   0   3       (@./kong/dynamic_hook/init.lua:227)
---- TRACE 1434 abort @./kong/dynamic_hook/init.lua:227 -- NYI: bytecode 71
```

Comment (bungle) [reply kong/dynamic_hook/init.lua]: @ADD-SP I did a small modification to this, not sure if it is effective.

**Review by ADD-SP — CHANGES_REQUESTED** [review]
Don't forget to correct the first argument like this: https://github.com/Kong/kong/blob/d43159afcec9212a4e214b7568c4922a832c1651/kong/init.lua#L325C1-L326C1

**Comment by bungle** [comment]
> Don't forget to remove the first argument at the definition of `run_hooks()`
> 
> https://github.com/Kong/kong/blob/d43159afcec9212a4e214b7568c4922a832c1651/kong/init.lua#L325

I have removed the `ctx` from `run_hooks`.

**Review by ADD-SP — APPROVED** [review]

**Review by samugi — APPROVED** [review]

**Pull request merged**